### PR TITLE
ui: Partitions Application Layer

### PR DESCRIPTION
**Pull request by johncowen**
TBD

~Kind of still in draft~, but good to start getting 👀 here, so definitely open for review as it stands. There will be some changes ontop. I'll add a description shortly (best viewing commit by commit for now).

### CODE REVIEW

File: ui/packages/consul-ui/app/instance-initializers/container.js
Comment (johncowen) [inline ui/packages/consul-ui/app/instance-initializers/container.js:16]: `ui-config` should be a repository. Added a little workaround here, rather than moving things around, but that should be changed as soon as possible to make it work like the rest of the application (including a completely async API)

File: ui/packages/consul-ui/app/adapters/partition.js
Comment (johncowen) [inline ui/packages/consul-ui/app/adapters/partition.js:12]: Its just easier to do all this (re-shaping the response and the request) in the same place.

The actual request is tightly coupled to the response, so the adapter/serializer split has always felt to me like decoupling for the sake of decoupling. Doing it here for all our adapters would also mean we can get rid of a bunch of code that is a little bit out of the ordinary (our fake HTTP headers to pass this value around), plus all the tests that go with it. Ideally we'd have an even nicer format to use for re-shaping but for now this is fine.

File: ui/packages/consul-ui/app/components/consul/acl/disabled/index.hbs
Comment (johncowen) [inline ui/packages/consul-ui/app/components/consul/acl/disabled/index.hbs:28]: We now have an 'ACLs disabled' component instead of building it into AppView. I've left the old on in AppView for the moment. It will probably be removed as part of the removal of legacy ACLs.

File: ui/packages/consul-ui/app/components/data-loader/index.hbs
Comment (johncowen) [inline ui/packages/consul-ui/app/components/data-loader/index.hbs:64]: If its a 401 (unauthorized) then don't show the disconnection message, just immediately try to reauth. Failure will then just show our default 403 error page.

File: ui/packages/consul-ui/app/components/data-source/index.js
Comment (johncowen) [inline ui/packages/consul-ui/app/components/data-source/index.js:99]: If the data is a single item and it hasn't been saved at all, then roll it back when the DataSource is removed from the page.

File: ui/packages/consul-ui/app/components/data-source/index.js
Comment (johncowen) [inline ui/packages/consul-ui/app/components/data-source/index.js:188]: Allows us to reload individual DataSources by pretending to remove them from the DOM and then pretending to re add them. We don't need/use this yet, but it will be handy later on. The FIXME here relates to not supporting `@loading="lazy"` DataSources yet. We can either do this here, or convert this to a TODO for merge. There's no rush to do this as we don't need it yet.

File: ui/packages/consul-ui/app/components/route/title/index.hbs
Comment (johncowen) [inline ui/packages/consul-ui/app/components/route/title/index.hbs:13]: The whitespace that is added by Handlebars when you do

```hbs
Navigated to {{@title}}
{{! yields (note the carriage return) }}
Navigated to
Title of page
```

...is noticeable when this is read out. Concat'ing this here works around that.

File: ui/packages/consul-ui/app/controllers/application.js
Comment (johncowen) [inline ui/packages/consul-ui/app/controllers/application.js:27]: We use the invalidate word in several other places in the codebase that all mean the same as what we are doing here - clearing the a cache. This is the only place we use it on the store so I decided to change it in this PR.

File: ui/packages/consul-ui/app/mixins/with-blocking-actions.js
Comment (johncowen) [inline ui/packages/consul-ui/app/mixins/with-blocking-actions.js:148]: I moved a few of our other mixins into here and deleted the old ones. All are mixins will be gone soon anyway, and they were originally only separate for organizations sake. Now we only have one `with-blocking-actions` Mixins left instead of a few.

File: ui/packages/consul-ui/app/routing/route.js
Comment (johncowen) [inline ui/packages/consul-ui/app/routing/route.js:147]: Make transitionTo and replaceWith available to the `route-action` helper, along with a micro-tick to avoid any ember recalculation/reevaluation console errors.

File: ui/packages/consul-ui/app/services/data-sink/protocols/http.js
Comment (johncowen) [inline ui/packages/consul-ui/app/services/data-sink/protocols/http.js:21]: DataSink will eventually use a `@dataSink` decorator just like `@dataSource` which will mean all of this can go.

File: ui/packages/consul-ui/app/services/repository/dc.js
Comment (johncowen) [inline ui/packages/consul-ui/app/services/repository/dc.js:25]: This should use proper errors instead of an object. I want to stop using the custom ember-data shaped errors as we don't need them. So ideally we'd just go back to using a standard shaped error.

File: ui/packages/consul-ui/app/services/repository/metrics.js
Comment (johncowen) [inline ui/packages/consul-ui/app/services/repository/metrics.js:27]: This needs to all be changed to use an asynchronous flow like all our other repositories. I figured I could do that later so explicitly called what it does now to `getSync` so its clear that it synchronous even though it should all be async. We can tackle this later on.

File: ui/packages/consul-ui/app/services/ui-config.js
Comment (johncowen) [inline ui/packages/consul-ui/app/services/ui-config.js:30]: All of this stuff is now so much easier to think about (the stuff we removed was in the application error hook)

**Review by johncowen — COMMENTED** [review]
Added a bunch of inline notes before adding a PR description

**Comment by jnwright** [comment]
@johncowen dropdown selector title should be singular (Admin Partition)

**Comment by johncowen** [comment]
@jnwright oh good spot! 😂 Removed, thankyou!

**Comment by johncowen** [comment]
Just a quick note here that this is all now passing green. There are a few FIXMEs I would like to get to, but again they shouldn't block an approval/merge. Gonna think about whether I want to do those in another PR 🤔

File: ui/packages/consul-ui/tests/acceptance/dc/acls/list-order.feature
Comment (johncowen) [inline ui/packages/consul-ui/tests/acceptance/dc/acls/list-order.feature:40]: Anything problematic to do with legacy ACLs I just deleted. It would be wasted work to make sure anything to do with legacy ACLs continued to pass/work seeing as we are removing legacy ACLs for 1.11.

I specifically only removed things that were problematic in some way and left the actual removal of legacy ACLs for a further PR.

File: ui/packages/consul-ui/tests/integration/adapters/kv-test.js
Comment (johncowen) [inline ui/packages/consul-ui/tests/integration/adapters/kv-test.js:155]: As explained up above, I added an `async` keyword to some adapter methods (but only the ones necessary for this PR). This meant that the mock things we create to test things also need to by 'promised'. So the majority of changes to the adapter tests here do exactly the same as this change, again and again and again.

File: ui/packages/consul-ui/tests/integration/serializers/auth-method-test.js
Comment (johncowen) [inline ui/packages/consul-ui/tests/integration/serializers/auth-method-test.js:87]: Changes to serializer tests just add `partition`, so again the changes here are done the same for every data model, again and again and again.

File: ui/packages/consul-ui/tests/integration/services/repository/auth-method-test.js
Comment (johncowen) [inline ui/packages/consul-ui/tests/integration/services/repository/auth-method-test.js:48]: Similar to serializer tests, repo test changes just add `partition` again and again and again for every single repository.

File: ui/packages/consul-ui/tests/steps/interactions/form.js
Comment (johncowen) [inline ui/packages/consul-ui/tests/steps/interactions/form.js:15]: As mentioned in the commit, for some reason ember-cli-page-object 'fillables' just stopped working, no idea why, I switched back to ember test helpers instead which I believe is more up to date than ember-cli-page-object things.

File: ui/packages/consul-ui/tests/unit/mixins/creating-route-test.js
Comment (johncowen) [inline ui/packages/consul-ui/tests/unit/mixins/creating-route-test.js:12]: With this change I could finally get rid of a load of our mixins, again this will help us get onto the next ember LTS.

File: ui/packages/consul-ui/tests/unit/routes/dc/intentions/edit-test.js
Comment (johncowen) [inline ui/packages/consul-ui/tests/unit/routes/dc/intentions/edit-test.js:11]: Most of our `Route` tests are removed as we no longer have Routes to test.

**Review by johncowen — COMMENTED** [review]
Added a bunch more inline comments.

**Review by kaxcode — CHANGES_REQUESTED** [review]
Rebasing needed before merge

**Comment by johncowen** [comment]
Thanks, I think this is good to go again

**Review by kaxcode — APPROVED** [review]
This PR is approved to be merged. Due to time constraints and the size of this PR, a review post-merge is required.

cc @mikemorris

**Comment by hc-github-team-consul-core** [comment]
:cherries: If backport labels were added before merging, cherry-picking will start automatically.

To retroactively trigger a backport after merging, add backport labels and re-run https://circleci.com/gh/hashicorp/consul/445517.

**Comment by johncowen** [comment]
Thanks @kaxcode , as I've just mentioned in the PR description here, please feel free to continue reviewing this here. Although it's merged here already, we can use this PR to get any of your questions and/or suggestions answered and make any further amends in the separate PR. In the meantime I'll be working on some of the FIXMEs for a separate PR. As discussed offline, I learnt a great approach from other folks here by using FIXMEs for things that should ideally by fixed in the PR itself, whilst TODOs are longer lived, so I'll be trying to figure those FIXMEs out until I hear back from you.

**Comment by kaxcode** [comment]
@johncowen - Post-merge review
Here are some questions that cover several files:
- Can you conscientiously explain what the most impactful changes were for adding Admin Partitions to the FE layer of the app?
- Why is `x-consul-index` header getting deleted in some adapters?
- Why are we creating a new ACLs disabled component instead of using the old one?
- How do the route.title and route.announcer changes relate to admin partitions?
- How do the changes to the `router.js` and all routes files relate to admin partitions?
- Why are the mixins getting deleted in this PR?
- Why are repo for nspace disabled and enabled deleted in this PR?

File: ui/packages/consul-ui/app/adapters/nspace.js
Comment (johncowen) [reply ui/packages/consul-ui/app/adapters/nspace.js:6]: Nspaces can now be different depending on which dc/partition you are in, so we need to qualify this call with an extra `dc` param.

File: ui/packages/consul-ui/app/components/consul/kv/form/index.hbs
Comment (johncowen) [reply ui/packages/consul-ui/app/components/consul/kv/form/index.hbs:23]: I think I changed this so that `parent` is the same as `parent.Key`. We only ever used `parent.Key` so figured it was easier to just use `parent` for a name.

File: ui/packages/consul-ui/app/components/data-form/index.js
Comment (johncowen) [reply ui/packages/consul-ui/app/components/data-form/index.js:49]: It was something related to changeset/ember-data. From what I remember I found that for some reason `isNew` wouldn't punch through the proxy to the underlying ember-data object, then I realised we have access here to the actual ember-data object a few lines up, so I just referenced that instead.

File: ui/packages/consul-ui/app/serializers/nspace.js
Comment (johncowen) [reply ui/packages/consul-ui/app/serializers/nspace.js:9]: I added `item.Namespace = item.Name` and also `item.Datacenter = query.dc`, for keying purposes. Also, now nspaces need to be keyed by DC, I moved it back to use the parent method so that it goes through the apps uid extra fingerprinting like the rest of the app (previously it didn't need the extra keying as namespaces used to be a top level bucket). The rest is just indenting.

File: ui/packages/consul-ui/app/services/repository/dc.js
Comment (johncowen) [reply ui/packages/consul-ui/app/services/repository/dc.js:34]: We don't uses these extra repo methods anymore. There's probably a bit more I can get rid of from this file actually 🤔 . The little refactor is just to add `partition` and use the parent class method instead of `store`.

File: ui/packages/consul-ui/app/services/repository/nspace.js
Comment (johncowen) [reply ui/packages/consul-ui/app/services/repository/nspace.js:7]: Its to make it all consistent (dc's, nspaces and partitions are all very similar now). TBH I think a lot of this can now be deleted 🤔

File: ui/packages/consul-ui/app/templates/application.hbs
Comment (johncowen) [reply ui/packages/consul-ui/app/templates/application.hbs:31]: I left quite a few comments in the file here to explain to folks what was happening. Its pretty much the same as what there used to be, just the code has been moved elsewhere. Is there anything specific here you want me to explain ontop of the comments?

**Comment by johncowen** [comment]
Hey! Thanks for the review and the questions!

I added a few inline answers and I'll try answer these other ones here:

> Can you consciously explain what the most impactful changes were for adding Admin Partitions to the FE layer of the app?

I'd say probably deleting a lot of code was the most impactful code change here. Means everything is a lot easier now. Namespace and partitions are now pretty much the same thing, and accessing params and data for them works just the same as any other param/data in the app. There are no little "oh, this is slightly different cos its nspaces" or "oh, you access this from here instead of where you usually would as its partitions". Now its all nice and consistent (mostly)

> Why is x-consul-index header getting deleted in some adapters?

This disables blocking queries for the 'section' so for example I think here specifically we use it to disable blocking queries for KV (which have never had blocking queries enabled)

> Why are we creating a new ACLs disabled component instead of using the old one?
There wasn't an old one. The only code for this was tied up in the AppView, which didn't make complete sense as the ACLs disabled view is only used in the ACLs area. Made sense to make it a component and only use it in ACLs.

> How do the route.title and route.announcer changes relate to admin partitions?
Previously the `@title` attribute meant we needed data available before the `Route` component was on the page:

```hbs
<Route @title={{item.Something}} />
```

...with this change we don't have some data until after the Route is on the page so we moved it to use a contextual component instead, so this change was required:

```hbs
<Route @title={{i can't access route here}} as |route|> <== I needed access to route
  <route.Title @title={{route.params}} /> <== now I have access to route and other things
</Route>
```

As I was doing this it felt like it was a good opportunity to make it more ergonomic and usable also, so I did that while I was there. I also took the opportunity to add some basic but useful documentation to explain this here https://github.com/hashicorp/consul/tree/581357c32a29de2640f2ec2f9b1274c4615db2f8/ui/packages/consul-ui/app/components/route#route hopefully that will also be useful moving forwards.

> How do the changes to the router.js and all routes files relate to admin partitions?

We don't need the majority of routes files anymore, there was a fair amount of logic duplication in them. Now there isn't, it's all in one place. There are super minimal changes to `router.js` related to whether ACLs are enabled or not.

> Why are the mixins getting deleted in this PR?

We don't need most if them anymore. I combined all the remaining ones into one file (which will soon be deleted so we can upgrade ember). There was a fair amount of legacy ACLs that was in the way a little here (we probably should have done that "Legacy ACL removal" work first). So I deleted quite a bit of stuff to avoid going through the extra effort it work for partitions, which would have been pointless.

> Why are repo for nspace disabled and enabled deleted in this PR?

This was originally a way to make it less complex to write differences between both 'modes'. Nspaces have been around a while now and I realised that the differences are so minimal this extra split into two classes was almost pointless, so I removed them here, this meant I could follow the same thing for the new partitions repository (which also doesn't have an enabled/disabled split).

Lemme know if you have any more q's.

**Pull request merged**
